frontend/javascripts/oxalis/view/action-bar/share_view_dataset_modal_view.js

```diff
@@ -0,0 +1,123 @@
+// @flow
+import { Divider, Modal, Input, Button, Row, Col } from "antd";
+import { connect } from "react-redux";
+import Clipboard from "clipboard-js";
+import React, { PureComponent } from "react";
+import type { APIDataset } from "admin/api_flow_types";
+import type { OxalisState } from "oxalis/store";
+import { getDatasetSharingToken } from "admin/admin_rest_api";
+import Toast from "libs/toast";
+import window from "libs/window";
+import _ from "lodash";
+
+type OwnProps = {|
+  isVisible: boolean,
+  onOk: () => void,
+|};
+type StateProps = {|
+  dataset: APIDataset,
+|};
+type Props = {| ...OwnProps, ...StateProps |};
+
+type State = {
+  sharingToken: string,
+};
+
+class ShareViewDatasetModalView extends PureComponent<Props, State> {
+  state = {
+    sharingToken: "",
+  };
+
+  componentDidMount() {
+    this.fetch();
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    if (
+      this.props.dataset.name !== prevProps.dataset.name ||
+      this.props.dataset.isPublic !== prevProps.dataset.isPublic ||
+      this.props.dataset.owningOrganization !== prevProps.dataset.owningOrganization
+    ) {
+      this.fetch();
+    }
+  }
+
+  async fetch() {
+    const { name, owningOrganization, isPublic } = this.props.dataset;
+    if (isPublic) {
+      // We only need to fetch a sharing token if the dataset is not public.
+      return;
+    }
+    const datasetId = { name, owningOrganization };
+    try {
+      const sharingToken = await getDatasetSharingToken(datasetId, { showErrorToast: false });
+      this.setState({ sharingToken });
+    } catch (error) {
+      console.error(error);
+    }
+  }
```

Comment:
I think, this is a good usecase for react hooks. Are you familiar with them? You could create a `useDatasetSharingToken` hook, which could then be used like this in the render component:

```
function ShareViewDatasetModalView({dataset}) {
  const datasetSharingToken = useDatasetSharingToken(dataset);
  ...
}
```

The benefit would be that the dataset sharing code could be refactored (via the hook) so that both sharing modals (tracing and dataset) can use this hook easily. The existing code in `componentDidUpdate` wouldn't be necessary, either, since the hook takes care of this.
The downside is that both modals would need to be refactored to be function components (opposed to class components). This is quite a bit of work (especially, re-hooking the components up to the store). So, I'm not really sure whether this is a good idea now. But if you feel like it you could familiarize yourself a bit with react hooks and fiddle around with it :)

Reply:
Honestly the moment I finished this component the idea of using hooks also came into my mind as the `ShareViewDatasetModal` component has little state behaviour. I already worked with react hooks in the past and using them for both modals is a good idea, in my opinion, to reduce the code redundancy and also improve the maintainability.

I think I'll refactor both components. Also to refresh my knowledge about hooks :+1:

Reply:
Awesome :+1: